mirrord/layer/src/file/hooks.rs

```diff
@@ -985,6 +985,84 @@ unsafe extern "C" fn realpath_darwin_extsn_detour(
 //     result
 // }
 
+/// Hook for `libc::readv`.
+#[hook_guard_fn]
+pub(crate) unsafe extern "C" fn readv_detour(
+    fd: RawFd,
+    iovecs: *const iovec,
+    iovec_count: c_int,
+) -> ssize_t {
+    if iovec_count < 0 {
+        return FN_READV(fd, iovecs, iovec_count);
+    }
+    let iovs = slice::from_raw_parts(iovecs, iovec_count as usize);
+    let read_size: u64 = iovs.iter().map(|iov| iov.iov_len as u64).sum();
+
+    read(fd, read_size)
+        .map(|read_file| {
+            let ReadFileResponse { bytes, read_amount } = read_file;
+
+            let mut copied = 0;
+            let mut iov_index = 0;
+
+            while copied < read_amount {
+                let iov = &iovs[iov_index];
+                let read_ptr = bytes.as_ptr();
+                let copy_amount = std::cmp::min(read_amount as usize, iov.iov_len);
+                let out_buffer = iov.iov_base.cast();
+                ptr::copy(read_ptr, out_buffer, copy_amount);
+                copied += copy_amount as u64;
+                // we trust iov_index to be in correct size since we checked it before
+                iov_index += 1;
+            }
+
+            // WARN: Must be careful when it comes to `EOF`, incorrect handling may appear as the
+            // `read` call being repeated.
+            ssize_t::try_from(read_amount).unwrap()
+        })
+        .unwrap_or_bypass_with(|_| FN_READV(fd, iovecs, iovec_count))
+}
+
+/// Hook for `libc::readv`.
+#[hook_guard_fn]
+pub(crate) unsafe extern "C" fn preadv_detour(
```

Comment:
It's a bit hard to read this with a bunch of github actions warnings in the middle, but, doesn't this duplicate a bunch of stuff from `readv`?

Doesn't this warrant the `hookx_logic` we have for shared implementations?

Reply:
it's not similar - but the copying to iovec is so I moved it into another function.

Reply:
Alright!